### PR TITLE
Add the boilerpipe parsing adapted from NUTCH-961

### PR DESCRIPTION
**Pull request by jeremie70**
I adapted a bit the patch from the [NUTCH-961](https://issues.apache.org/jira/browse/NUTCH-961). This patch add the boilerpipe parser in nutch which improve a lot the quality of parser.

### CODE REVIEW

File: conf/nutch-default.xml [hunk 1]
```diff
@@ -876,6 +876,19 @@
   </description>
 </property>
 
+<!-- tika properties -->
+
+<property>
+  <name>tika.boilerpipe</name>
+  <value>false</value>
```
Comment (lewismc) [inline conf/nutch-default.xml]: Can you provide descriptions of these properties please?

File: src/plugin/parse-tika/src/java/org/apache/nutch/parse/tika/BoilerpipeExtractorRepository.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nutch.parse.tika;
+
+import java.lang.ClassLoader;
+import java.lang.InstantiationException;
+import java.util.WeakHashMap;
+import org.apache.commons.logging.Log;
```
Comment (lewismc) [inline src/plugin/parse-tika/src/java/org/apache/nutch/parse/tika/BoilerpipeExtractorRepository.java]: Nutch currently uses Slf4j

org.slf4j.Logger
org.slf4j.LoggerFactory

I think!

File: src/plugin/parse-tika/src/java/org/apache/nutch/parse/tika/BoilerpipeExtractorRepository.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        // Attempt to load the class
+        try {
+          ClassLoader loader = BoilerpipeExtractor.class.getClassLoader();
+          Class extractorClass = loader.loadClass(boilerpipeExtractorName);
+
+          // Add an instance to the repository
+          extractorRepository.put(boilerpipeExtractorName, (BoilerpipeExtractor)extractorClass.newInstance());
+
+        } catch (ClassNotFoundException e) {
+          LOG.error("BoilerpipeExtractor " + boilerpipeExtractorName + " not found!");
```
Comment (lewismc) [inline src/plugin/parse-tika/src/java/org/apache/nutch/parse/tika/BoilerpipeExtractorRepository.java]: In slf4j we can better structure the catch
http://www.slf4j.org/faq.html#logging_performance
e.g.
```
logger.debug("The entry is {}.", entry);
```

File: src/plugin/parse-tika/src/java/org/apache/nutch/parse/tika/TikaParser.java [hunk 4]
```diff
@@ -109,7 +114,18 @@ public Parse getParse(String url, WebPage page) {
     HTMLDocumentImpl doc = new HTMLDocumentImpl();
     doc.setErrorChecking(false);
     DocumentFragment root = doc.createDocumentFragment();
-    DOMBuilder domhandler = new DOMBuilder(doc, root);
+   // DOMBuilder domhandler = new DOMBuilder(doc, root);
+    ContentHandler domHandler;
+    // Check whether to use Tika's BoilerplateContentHandler
+    if (useBoilerpipe) {
+        LOG.debug("Using Tikas's Boilerpipe with Extractor: " + boilerpipeExtractorName);
```
Comment (lewismc) [inline src/plugin/parse-tika/src/java/org/apache/nutch/parse/tika/TikaParser.java]: Can also use more efficient slf4j convention
logger.debug("The entry is {}.", entry);

**Comment by lewismc** [comment]
Nice patch. Do you have any test cases in mind?

**Comment by jeremie70** [comment]
Thank you a lot for your help. 

I didn't know that this enhancement of theses logs can improve the performance more than 30 times. It's impressive. :)
If you want test it, you have to give theses properties to nutch-site.xml : 
```
<property>
  <name>http.content.limit</name>
  <value>-1</value>
</property>

<property>
  <name>tika.boilerpipe</name>
  <value>true</value>
</property>

<property>
  <name>plugin.includes</name>
 <value>protocol-http|urlfilter-regex|parse-(tika)|index-(basic|anchor)|urlnormalizer-(pass|regex|basic)|scoring-opic</value>
</property>
```
After a recompile you can try and compare with or without tika.boilerpipe : 

> ./nutch indexchecker URL

**Comment by chrismattmann** [comment]
@jeremie70 can you please update this PR? I'll go ahead and get it committed or @lewismc can

**Comment by jeremie70** [comment]
I did it. I forgot something ?

**Comment by chrismattmann** [comment]
i think you did forget something since I don't see any changes from you here on this PR since @lewismc comments. Did you make sure to push to your remote so this PR gets updated?

**Comment by jeremie70** [comment]
I'm a new user that is possible that i missed something. 
But i pushed a new commit with descriptions in nutch-default.xml, i modified the system of log and i added a property in the indexchecker to see all the field. It's appear in the "files changed" so i think that i updated the patch. Can you check again @chrismattmann  please ? Check if you see a description below the value of :
```
<property>
  <name>tika.boilerpipe</name>
  <value>false</value>
```

**Comment by lewismc** [comment]
@jeremie70 no nothing has been updated. Can you please push your branch (say your branch was origin), something like
```
git push origin my-branch
```
Thanks

**Comment by jeremie70** [comment]
I found the problem, i pushed with the bad email, so github didn't show to other my commit. 
I will resolve this tomorow.
Thank you for your help. ;)

**Comment by jeremie70** [comment]
It would be good now. :)

**Comment by chrismattmann** [comment]
thanks @jeremie70 i will review cc @lewismc

**Comment by chrismattmann** [comment]
merged thanks @jeremie70

**Comment by dastergon** [comment]
In which Nutch 2.x version it will land?

**Comment by lewismc** [comment]
@dastergon the very next release which is currently slated for 2.4

**Comment by dastergon** [comment]
@lewismc is there any ETA for that version?

**Comment by lewismc** [comment]
If you can please help me with https://github.com/apache/nutch/pull/99 it would be greatly appreciated.
This is the only blocker.

**Comment by dastergon** [comment]
@lewismc ok, I will try to reproduce it

**Pull request merged**